apps/pushy/src/pushy_principal.erl

```diff
@@ -41,56 +41,74 @@ fetch_principal(OrgName, Requestor) ->
-               {"X-Ops-Content-Hash", ""},
-               {"X-Ops-Sign", ""},
-               {"X-Ops-Timestamp", ""},
-               {"X-Chef-Version", ChefApiVersion}],
-    Url = api_url(OrgName, Requestor),
-    case ibrowse:send_req(Url, Headers, get) of
-        {ok, "404", _Headers, ResponseBody} ->
-            %% A 404 can mean either the org doesn't exist, or the principal doesn't (or
-            %% isn't in the given org). We need to parse the JSON response body to find out
-            %% which is the case.
-            EJson = jiffy:decode(ResponseBody),
-            Reason = case ej:get({"not_found"}, EJson) of
-                         <<"org">> -> org;
-                         <<"principal">> -> principal
-                     end,
-            throw({error, {not_found, Reason}});
-        {ok, Code, ResponseHeaders, ResponseBody} ->
-            ok = pushy_http_common:check_http_response(Code, ResponseHeaders,
-                                                       ResponseBody),
-            parse_json_response(ResponseBody);
+    Path = api_path(OrgName, Requestor),
+    Response = pushy_http_common:fetch_unauthenticated(Path),
+    case parse_principal_response(Response) of
         {error, Reason} ->
-            io:format("got error ~p~n", [Reason]),
-            throw({error, Reason})
+            lager:error("Error requesting principal ~p: ~p~n", [Requestor, Reason]),
+            throw({error, Reason});
+        [#pushy_principal{}|_Rest] = Principals ->
+            Principals
     end.
 
--spec parse_json_response(Body :: binary()) -> #pushy_principal{}.
-parse_json_response(Body) ->
-    EJson = jiffy:decode(Body),
+parse_principal_response({error, Reason}) ->
+    {error, Reason};
+parse_principal_response(#chef_api_response{response_code = "404",
+                                            response_body = Body
+                                           }) ->
+    Reason = case ej:get({"not_found"}, Body) of
+                 <<"org">> -> org;
+                 <<"principal">> -> principal
+             end,
+    {error, {not_found, Reason}};
+parse_principal_response(#chef_api_response{response_api_version = 1,
+                                            response_body = EJson} = Resp) ->
+    case pushy_http_common:check_response(Resp) of
+        ok ->
+            Principals = ej:get({"principals"}, EJson),
+            Principals1 = [parse_ejson_response(P) || P <- Principals],
+            filter_not_associated(Principals1);
+        {error, Reason} ->
+            {error, Reason}
+    end;
+parse_principal_response(#chef_api_response{response_api_version = 0,
+                                            response_body=EJson} = Resp) ->
+    case pushy_http_common:check_response(Resp) of
+        ok ->
+            filter_not_associated([parse_ejson_response(EJson)]);
+        {error, Reason} ->
+            {error, Reason}
+    end.
+
+filter_not_associated(Principals) ->
```

Comment:
Perhaps a comment here on what this function is doing. I get it by reading all of it, but... I read all of it :)